layers/+lang/haskell/README.org

```diff
@@ -116,28 +134,48 @@ Backend can be chosen on a per project basis using directory local variables
 
 *Note:* you can easily add a directory local variable with ~SPC f v d~.
 
+There are two backends available: =dante= and =lsp=.
+
+=dante= is lightweight, requires no setup and works out of the box in most cases, which is why it is also a default backend.
+
+=lsp= (=hie= or =hls=) is a more ambitious, heavy-weight, cutting-edge backend that is however still somewhat rough on the edges and requires some additional setup.
+
 *** =dante=
-[[https://github.com/jyp/dante][Dante]] is a fork of Intero mode which aims exclusively at providing a convenient frontend to GHCi.
+[[https://github.com/jyp/dante][Dante]] is a lightweight backend which delegates most of its work directly to GHCi.
 
 It brings features like syntax checking, auto completion, hlint suggestions, automatic error fixing, info at point, definition and use sites.
 
 =dante= works for =cabal=, =nix=, =sytx=, and =stack= users and requires no additional setup.
 
-=dante= requires Emacs 25.
-
 *** =lsp=
-=lsp= requires an appropriate installation of =hie=, the official Haskell language server, to work properly.
+[[https://microsoft.github.io/language-server-protocol][Language Server Protocol]] is a standard for implementing language backends.
+
+In Haskell layer, you can use a backend that implements Language Server Protocol for Haskell by specifying =lsp= as backend
+and then installing concrete backend implementation, of which there are two available at the moment: =hie= and =hls=.
+
+Enabling the =lsp= backend requires the =lsp= layer to be enabled, and provides access to
+all the additional =lsp-mode= key bindings. =lsp= uses =lsp-ui= to show the errors/warnings instead of =flycheck=. 
```

Comment:
Well the line about LSP not using flycheck is not right, actually there is a specific checker registered within the flycheck system which nowadays is called `lsp` not `lsp-ui` anymore (latest version). However that's still the flycheck system so the same setup instructions apply.

To elaborate a bit more, the traditional emacs system is to register multiple checkers with flycheck which then run sequentially. This is normally slow and requires a lot of setup within emacs. Lsp-mode in contrary passes the job of maintaining linter threads to the external LSP server allowing them to run in parallel. For the emacs user this is visible by that there is only one checker registered called `lsp`. Normally this is ok, however if the server is not well enough designed to be able to run multiple linters some layers have tried to register a second checker after `lsp` this however is very difficult to get right and again is very slow. The right approach in this case is to open a PR for the LSP server to support that specific linter.

Reply:
Thanks for detailed explanation! But from what I read / understood, LSP errors don't appear in flycheck's error list? Or do they then? I would try it myself but I don't have HIE/HLS set up at the moment, might get to it during the weekend but certainly not sooner.

Reply:
Errors from lsp are registered as flycheck errors, as are all errors you see in any buffer (only if it is using very obscure or veeeerrryyy old code for this).